src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -36,21 +36,22 @@ public abstract class PackageReferenceProject<T, U> : BuildIntegratedNuGetProjec
     {
         internal static readonly Comparer<PackageReference> PackageReferenceMergeComparer = Comparer<PackageReference>.Create((a, b) => a?.PackageIdentity?.CompareTo(b.PackageIdentity) ?? 1);
 
-        private protected readonly Dictionary<string, TransitiveEntry> TransitiveOriginsCache = new();
-
         private readonly protected string _projectName;
         private readonly protected string _projectUniqueName;
         private readonly protected string _projectFullPath;
 
         // Cache
+        private protected Dictionary<string, TransitiveEntry> TransitiveOriginsCache { get; set; }
         protected T InstalledPackages { get; set; }
         protected T TransitivePackages { get; set; }
 
-        private readonly object _lockObj = new object();
+        private readonly object _installedAndTransitivePackagesLock = new object();
+        private readonly object _transitiveOriginsLock = new object();
 
         private protected DateTime _lastTimeAssetsModified;
         private protected WeakReference<PackageSpec> _lastPackageSpec;
         private protected IList<LockFileItem> _packageFolders;
+        private readonly NuGetFrameworkSorter _frameworkSorter;
```

Comment:
If this is getting re-used by different threads, then it must be stateless and threadsafe. In this case, you can also make it static, so if the solution has 100 different PackageReferenceProjects, it doesn't need 100 instances of the sorter.  Another option is to add a static `NuGetFrameworkSorter.Instance`, and use that in this code, since different classes using this object don't need different static instances.

Reply:
Great points. I'll make a static instance since it's stateless. That way only one instance will be allocated regardless of how many PackageReferenceProjects exist. It also is consistent with the PackageReferenceMergeComparer which is also static.